clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.kafka.common.security.scram.ScramMessages.ServerFirstMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SaslClient implementation for SASL/SCRAM.
+ * <p>
+ * This implementation expects a login module that populates username as
+ * the Subject's public credential and password as the private credential.
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc5802">RFC 5802</a>
+ *
+ */
+public class ScramSaslClient implements SaslClient {
+
+    private static final Logger log = LoggerFactory.getLogger(ScramSaslClient.class);
+
+    enum State {
+        SEND_CLIENT_FIRST_MESSAGE,
+        RECEIVE_SERVER_FIRST_MESSAGE,
+        RECEIVE_SERVER_FINAL_MESSAGE,
+        COMPLETE,
+        FAILED
+    };
+
+    private final ScramMechanism mechanism;
+    private final CallbackHandler callbackHandler;
+    private final ScramFormatter formatter;
+    private String clientNonce;
+    private State state;
+    private byte[] saltedPassword;
+    private ScramMessages.ClientFirstMessage clientFirstMessage;
+    private ScramMessages.ServerFirstMessage serverFirstMessage;
+    private ScramMessages.ClientFinalMessage clientFinalMessage;
+
+    public ScramSaslClient(ScramMechanism mechanism, String authorizationId, CallbackHandler cbh) throws NoSuchAlgorithmException {
+        this.mechanism = mechanism;
+        this.callbackHandler = cbh;
+        this.formatter = new ScramFormatter(mechanism);
+        setState(State.SEND_CLIENT_FIRST_MESSAGE);
+    }
+
+    @Override
+    public String getMechanismName() {
+        return mechanism.mechanismName();
+    }
+
+    @Override
+    public boolean hasInitialResponse() {
+        return true;
+    }
+
+    @Override
+    public byte[] evaluateChallenge(byte[] challenge) throws SaslException {
+        switch (state) {
+            case SEND_CLIENT_FIRST_MESSAGE:
+                if (challenge != null && challenge.length != 0)
+                    throw new SaslException("Expected empty challenge");
+                clientNonce = formatter.secureRandomString();
+                NameCallback nameCallback = new NameCallback("Name:");
+                try {
+                    callbackHandler.handle(new Callback[]{nameCallback});
+                } catch (IOException | UnsupportedCallbackException e) {
+                    setState(State.FAILED);
+                    throw new SaslException("User name could not be obtained", e);
+                }
+                String username = nameCallback.getName();
+                String saslName = formatter.saslName(username);
+                this.clientFirstMessage = new ScramMessages.ClientFirstMessage(saslName, clientNonce);
+                setState(State.RECEIVE_SERVER_FIRST_MESSAGE);
+                return clientFirstMessage.toBytes();
+
+            case RECEIVE_SERVER_FIRST_MESSAGE:
+                this.serverFirstMessage = new ServerFirstMessage(challenge);
```

Comment:
Calls like this could generate SaslException. Should we catch those, set the state to fail and rethrow?